### PR TITLE
Multiple `__LINE__` macro detection fixes and improvements

### PR DESCRIPTION
**Pull request by jpoimboe**
Fix a couple of issues reported in #1253.

### CODE REVIEW

**Comment by jpoimboe** [comment]
Going to rework the "line macro change only" patch a bit.  In the process I found some bugs in the ppc-specific version of the function.  Also there's a lot of copy/paste between the x86 and ppc versions which I'll just converge into a single function.

**Comment by jpoimboe** [comment]
> 1. I'm pretty sure there is an endian bug in the ppc64le version (it compares whole word instruction encoding values, which wouldn't work in cross-arch cases)

Is this only seen when running ppc64le unit tests on an s390 host?  I hadn't really considered that scenario.

> 2. Take a look at Erastus's WIP [s390x version](https://github.com/dynup/kpatch/pull/1243/commits/da775bfe1906916e5ecfab30e3a37244f8079118), it should be very easy to fold into a common function along with the other arches.

What exactly should I look at?  It looks really similar to the ppc and x86 versions..  What I was thinking was the main differences between the arch implementations could be abstracted with a few functions like 'insn_length()' and `is_potential_line_immedate_load_insn()` or such.

**Comment by jpoimboe** [comment]
v2:
- unify kpatch_line_macro_change_only() for all arches
- check for both `__func__` and `__btrfs_abort_transaction`

**Comment by joe-lawrence** [comment]
> > 1. I'm pretty sure there is an endian bug in the ppc64le version (it compares whole word instruction encoding values, which wouldn't work in cross-arch cases)
> 
> Is this only seen when running ppc64le unit tests on an s390 host? I hadn't really considered that scenario.
> 

Yes, I think that would do it.  An odd scenario for sure, but I think checking per-byte would be safer and avoid anyone else from copy/pasting that pattern in the future.  (I think Erastus and I hit this on x86 when we were testing s390 object files and modeled after the ppc64le implementation of checking long int values.)

> > 2. Take a look at Erastus's WIP [s390x version](https://github.com/dynup/kpatch/pull/1243/commits/da775bfe1906916e5ecfab30e3a37244f8079118), it should be very easy to fold into a common function along with the other arches.
> 
> What exactly should I look at? It looks really similar to the ppc and x86 versions.. What I was thinking was the main differences between the arch implementations could be abstracted with a few functions like `insn_length()` and `is_potential_line_immediate_load_insn()` or such.

Sorry to be vague, but yeah that's pretty much it.  If you were creating a common function, I just wanted to give you a heads up that the pattern pretty much holds for s390, too.

File: kpatch-build/create-diff-object.c [hunk 1]
```diff
@@ -623,153 +680,86 @@ static bool kpatch_line_macro_change_only_x86_64(struct section *sec)
...
-		/* verify it's a load immediate to r5 */
-		if (!(instr1 == 0x38a0 && instr2 == 0x38a0))
-			return false;
+			if (r->offset < offset + insn1_len)
+				continue;
 
-		found = 0;
-		list_for_each_entry(rela, &sec->rela->relas, list) {
-			if (rela->offset < offset + PPC_INSTR_LEN)
+			rela = toc_rela(r);
```
Comment (joe-lawrence) [inline kpatch-build/create-diff-object.c:725]: Ok, **toc_rela()** should just return the **rela** back for non-ppc64le.

File: kpatch-build/create-diff-object.c [hunk 2]
```diff
@@ -577,14 +577,76 @@ static void kpatch_compare_correlated_section(struct section *sec)
...
+	struct insn insn;
+	unsigned int instr;
+
+	switch(kelf->arch) {
+
+	case X86_64:
+		insn_init(&insn, addr, 1);
+		insn_get_opcode(&insn);
+
+		if (insn.opcode.value == 0xba)	/* mov $imm, %edx */
```
Comment (joe-lawrence) [inline kpatch-build/create-diff-object.c:619]: Looks like there is a union aliasing **value** and a character array of **bytes**:
``` C
struct insn_field {
        union {
                insn_value_t value;
                insn_byte_t bytes[4];
        };
```
I think checking specific **bytes[x]** would be endian safer.

File: kpatch-build/create-diff-object.c (hunk 1 above)
Comment (jpoimboe) [reply kpatch-build/create-diff-object.c:725]: Right, and also for the ppc64le case where the rela isn't a toc-style nested rela.

File: kpatch-build/create-diff-object.c [hunk 3]
```diff
@@ -577,14 +577,76 @@ static void kpatch_compare_correlated_section(struct section *sec)
...
+			return true;
+		if (insn.opcode.value == 0xbe)	/* mov $imm, %esi */
+			return true;
+
+		return false;
+
+	case PPC64:
+		instr = *(unsigned int *)addr >> 16;
+
+		if (instr == 0x38a0)		/* li r5, imm */
```
Comment (joe-lawrence) [inline kpatch-build/create-diff-object.c:629]: Likewise here, **instr** is a native numerical (endian) type.  Something like this should work regardless of host/file endianness:
``` C
char *instr = addr + PPC_RA_OFFSET / 8;
if (instr[0] == 0xa0 || instr[1] == 0x38)
```

**Review by joe-lawrence — COMMENTED** [review]
This is a nice cleanup (and fixes an outstanding issue).

I left two comments about endianness.  If you want to defer that work until we do a fuller audit of the code when s390x support goes live, we can punt it down the road.  I happened to have a half-baked local branch that started addressing a few spots like this.

**Comment by jpoimboe** [comment]
> This is a nice cleanup (and fixes an outstanding issue).

Oh it does fix an issue, doesn't it ;-)  I assume you mean the toc_rela() change, I should move that to its own patch.  I'll also try to fix the endian issues in another patch.

**Comment by jpoimboe** [comment]
v3:
- Split toc_rela() change into a separate patch

(will also do a v4 with endian fixes)

**Comment by jpoimboe** [comment]
v4:
- Fix endian issues
- Add theoretical s390 support for `__LINE__` macro detection.  This is a modified version of the patch in #1243.  I went ahead and included this patch here because it helped me test the endian fixes, and it works fine as a standalone patch which can be easily combined with #1203 for further s390 testing going forward.

File: kpatch-build/create-diff-object.c [hunk 4]
```diff
@@ -742,7 +743,8 @@ static bool kpatch_line_macro_change_only(struct kpatch_elf *kelf,
 			    !strcmp(rela->sym->name, "___might_sleep") ||
 			    !strcmp(rela->sym->name, "__might_fault") ||
 			    !strcmp(rela->sym->name, "printk") ||
-			    !strcmp(rela->sym->name, "lockdep_rcu_suspicious")) {
+			    !strcmp(rela->sym->name, "lockdep_rcu_suspicious") ||
+			    !strcmp(rela->sym->name, "__btrfs_abort_transaction")) {
```
Comment (joe-lawrence) [inline kpatch-build/create-diff-object.c:769]: **fs/btrfs/ctree.h** also contains **btrfs_handle_fs_error** and **btrfs_panic** macros that similarly pass `__LINE__` to a function as the third parameter.  The latter two count up to the same number of affected .c files as the first, would they be worth handling, too?
```
$ grep -R -o -e btrfs_abort_transaction fs/btrfs | sort -u | wc -l
21
$ grep -R -o -e btrfs_handle_fs_error fs/btrfs | sort -u | wc -l
15
$ grep -R -o -e btrfs_panic fs/btrfs | sort -u | wc -l
5
```

Comment (jpoimboe) [reply kpatch-build/create-diff-object.c:769]: Good call, I'll add `__btrfs_handle_fs_error` and `__btrfs_panic`.

**Comment by jpoimboe** [comment]
v5:
- added `__btrfs_handle_fs_error` and `__btrfs_panic`

**Comment by jpoimboe** [comment]
v6:
- check arg2/arg3 on s390/ppc64le for parity with x86
- add "_printk" check (which was renamed from "printk" on newer kernels)

**Comment by jpoimboe** [comment]
Bumped unit test submodule reference

**Pull request merged**